### ISSUE TITLE
Add Movie does not set NumberAvailable

### ISSUE THREAD

**Issue by skalkan** [mosh-hamedani/vidly-mvc-5#1]
**NumberAvailable** column of movie table is not set anywhere in the code so when a new movie inserted it remains **null**. Nobody can see the newly added movies, unless he or she manually updates database and set NumberAvailable to a number greater than zero.

By the way,  thanks for this wonderful and awesome course and project.

**Comment by edwin4400** [comment]
Add the 2 lines in bold, should be able to set the NumberAvailable = NumberInStock.

```C#        
        [HttpPost]
        [ValidateAntiForgeryToken]
        public ActionResult Save(Movie movie)
        {
            if (!ModelState.IsValid)
            {
                var viewModel = new MovieFormViewModel(movie)
                {
                    Genres = _context.Genres.ToList(),
                    
                };

                return View("MovieForm", viewModel);
            }
            if (movie.Id == 0)
            {
                                                    movie.NumberAvailable = movie.NumberInStock;
                _context.Movies.Add(movie); //add to dbContext
            }
            else
            {
                                                    movie.NumberAvailable = movie.NumberInStock;
                var movieInDb = _context.Movies.Single(m => m.Id == movie.Id);
                movieInDb.Name = movie.Name;
                movieInDb.ReleaseDate = movie.ReleaseDate;
                movieInDb.GenreId = movie.GenreId;
                movieInDb.NumberInStock = movie.NumberInStock;
                movieInDb.NumberAvailable = movie.NumberAvailable;

            }
            
            _context.SaveChanges();

            return RedirectToAction("Index", "Movies");
        }

**Comment by nerdyrasa** [comment]
For the case where you're editing number in stock, you might want to take into account that some movies may have already been rented. This is how I handled that:

if (movie.Id == 0)
            {
                movie.DateAdded = DateTime.Now;
                movie.NumberAvailable = movie.NumberInStock;
                _context.Movies.Add(movie);
            }
            else
            {
                var movieInDb = _context.Movies.Single(m => m.Id == movie.Id);
                movieInDb.Name = movie.Name;
                movieInDb.GenreId = movie.GenreId;
                byte numberInStockNew = movie.NumberInStock;
                byte numberInStockOld = movieInDb.NumberInStock;
                byte moviesNetChange = (byte)(numberInStockNew - numberInStockOld);
                movieInDb.NumberAvailable += moviesNetChange;
                movieInDb.NumberInStock = numberInStockNew;
                movieInDb.ReleaseDate = movie.ReleaseDate;
            }

Also, we have similar code in the api for MoviesController that should really be updated to assign a value for NumberAvailable. We're using a dto and mapping, so I'm not sure how that would work. Any ideas?

**Comment by edwin4400** [comment]
Hi Nerdyrasa,

Shouldn't the "movieInDb.NumberInStock = numberInStockNew;" be
"movieInDb.NumberInStock += moviesNetChange;"

**Comment by nerdyrasa** [comment]
I don't think so. Say the number in stock is 15, then you change it to 18 using the edit form. You can overwrite it to 18 no problem since this is a fixed value--there are no calculations in the code for it. 

But the number available varies depending on rentals--you are decrementing this value every time it's rented. If the number available is 10 in the database since 5 were rented, you wouldn't want to just set the value to 18, but capture that 3 were added so 10+3 or 13 are now available. 

Does that make sense?

**Comment by nerdyrasa** [comment]
Actually, it would probably be best to keep track of number rented instead of number available. Number rented would initialize to 0 when the movie is created. If the number of stock is edited, number rented does not need updating. Number available can always be calculated from number in stock minus number rented.  Does that make sense?
